socketioxide/src/socket.rs

```diff
@@ -36,10 +38,119 @@ pub struct Socket<A: Adapter> {
     message_handlers: RwLock<HashMap<String, BoxedHandler<A>>>,
     ack_message: RwLock<HashMap<i64, oneshot::Sender<AckResponse<Value>>>>,
     ack_counter: AtomicI64,
-    tx: tokio::sync::mpsc::Sender<EnginePacket>,
     pub handshake: Handshake,
     pub sid: Sid,
     pub extensions: Extensions,
+    sender: Mutex<PacketSender>,
+}
+
+struct PacketSender {
+    tx: tokio::sync::mpsc::Sender<EnginePacket>,
+    bin_payloads: Option<VecDeque<EnginePacket>>,
+}
+
+impl PacketSender {
+    fn new(
+        tx: tokio::sync::mpsc::Sender<EnginePacket>,
+        failed_buffer: VecDeque<EnginePacket>,
+    ) -> Self {
+        Self {
+            tx,
+            bin_payloads: (!failed_buffer.is_empty()).then(|| failed_buffer),
+        }
+    }
+
+    fn send_raw(&mut self, mut packet: RetryablePacket) -> Result<(), TransportError> {
+        if let Err(err) = self.send_binaries() {
+            match err {
+                TransportError::SendFailedBinPayloads(None) => {}
+                TransportError::SocketClosed => return Err(TransportError::SocketClosed),
+                _ => unreachable!(),
+            };
+            Err(TransportError::SendMainPacket(packet))
+        } else {
+            let main_packet = packet.main_packet.take().unwrap();
+            match self.tx.try_send(main_packet) {
+                Err(TrySendError::Full(main_packet)) => {
+                    packet.main_packet = Some(main_packet);
+                    Err(TransportError::SendMainPacket(packet))
+                }
+                Err(TrySendError::Closed(_)) => Err(TransportError::SocketClosed),
+                _ => {
```

Comment:
Just put `Ok(())` rather than _ to avoid any future case mismatch and to understand the variant